### DISCUSSION TITLE
Vrp problem with compact routes

### DISCUSSION THREAD

**Discussion by pmtmd** [google/or-tools#3689] [Routing (and legacy CP) questions]
Hi everyone,

I'm working on a vehicle routing problem (VRP) using OR-Tools and I'm trying to create compact routes where the nodes are as close to each other as possible. I'd like to add a variable value that measures the compactness of each route to the objective function.

My problem is that the OR-Tools transit callback gets only two nodes as parameters, whereas I need to calculate the objective based on more than two nodes in each route. Is there a way or an approach to achieve this in OR-Tools?

Like fixedcostforallvehicles, i would like to add a variable value that represents the compactness of each route. Is there a way to achieve this using OR-Tools and if so, how?

I appreciate any help or insights on this topic.

Thank you in advance!

**Comment by Regista6** [comment]
This seems relevant, especially the quadratic cost factor. 
https://github.com/google/or-tools/blob/87a9dc8c3392e48597ac5b19ca46604791021330/ortools/constraint_solver/routing.h#L1088-L1104

**Comment by jmarca** [comment]
As far as I can recall, the objective function for the routing solver is pretty inaccessible. The only hooks that I know of are the soft constraints for upper and lower bounds. You might be able to hack up something using that. CPSat is more flexible in this regard, so you might consider rewriting everything and using the circuit constraint in place of the routing solver. But that is a lot of work. Thinking as I peck at my phone— This is a clustering problem of sorts, right? The routing minimization gives point to point minima, but you need some other metric. The shortest route might be a straight line. 

How much can be done in advance prior to solving?  What I’m thinking is: say you have 5 (or N) vehicles, so you need that many routes. Pick N equidistant locations that lie on the outer envelope of some boundary of your locations. Then set a second dimension, unique to each vehicle i, where distance is defined as distance to the ith point on the boundary. Then set a soft upper bound of zero for this secondary dimension. 

It might work that vehicle 1 will see points closest to point 1, veh 2 gets points closer to pont 2, and so on. That might be enough clustering for your needs. 

You also might be able to just straight up run a good k-means clustering algorithm and use the centroids of those clusters as the per vehicle points. That should work better, if the data are boring and sedate. But a k-means won’t work if there are operating hours differences, per-vehicle capacity quirks, or other issues. 

You also might consider using a third dimension from a central point, and call distance the angle from horizontal. Hmm no that won’t work because I can’t see how to translate that to an upper or lower bound test will work with the soft bound constraint API calls. Hope that helps or gives you some better ideas. 

Finally, if you cross-post, please say so

James
